<div style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h1 style = "padding: 15px; 
                 color: White;
                 text-align:center;
                 font-family: Trebuchet MS;">Netflix Movies and TV Shows
    </h1>
</div>

<div align="center"> 
    <img src="https://content.api.news/v3/images/bin/3fcf3cdcc2cfd6905e582a186e46e584" alt="Netflix" width="75%" style="margin-top:2rem;margin-bottom:2rem">
</div>

<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Netflix is one of the most popular media and video streaming platforms. They have over 8000 movies or tv shows available on their platform, as of mid-2021, they have over 200M Subscribers globally.</p>
    </div>
</body>

<div style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">Table of Contents
    </h2>
</div>

* [<span style="font-family: Trebuchet MS; font-size:15px;">1. Imports</span>](#imports)
* [<span style="font-family: Trebuchet MS; font-size:15px;">2. Loading the Dataset</span>](#loading-the-dataset)
* [<span style="font-family: Trebuchet MS; font-size:15px;">3. Understanding the Data</span>](#understanding-the-data)
    * [<span style="font-family: Trebuchet MS; font-size:15px;">3.1 Checking Null Values</span>](#checking-null-values)
    * [<span style="font-family: Trebuchet MS; font-size:15px;">3.2 Optimizing the Dataset</span>](#optimizing-the-dataset)
* [<span style="font-family: Trebuchet MS; font-size:15px;">4. Exploratory Data Analysis</span>](#exploratory-data-analysis)


<div id="imports"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">1. Imports
    </h2>
</div>

In [1]:
%%capture
import numpy as np
import pandas as pd
import plotly.express as px

<div id="loading-the-dataset"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">2. Loading the Dataset
    </h2>
</div>

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/sahinozan/Netflix-Movies-TV-Shows/main/netflix_titles.csv", encoding="utf-8")
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

<div id="understanding-the-data"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">3. Understanding the Data
    </h2>
</div>

<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p style="font-family:verdana;">Dataset consists of 12 features.</p>
        <ol>
            <li style="font-family:verdana;"><b>show_id</b>: The unique id number of each show or movie
                <ul>
                    <li style="font-family:verdana;">s1</li>
                    <li style="font-family:verdana;">s2</li>
                    <li style="font-family:verdana;">s3</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>type</b>: Type of the content
                <ul>
                    <li style="font-family:verdana;">Movie</li>
                    <li style="font-family:verdana;">TV Show</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>title</b>: Title of the content
                <ul>
                    <li style="font-family:verdana;">The Starling</li>
                    <li style="font-family:verdana;">Squid Game</li>
                    <li style="font-family:verdana;">Jaws: The Revenge</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>director</b>: Name of the director for that movie or show
                <ul>
                    <li style="font-family:verdana;">Steven Spielberg</li>
                    <li style="font-family:verdana;">Cédric Jimenez</li>
                    <li style="font-family:verdana;">Hirotsugu Kawasaki</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>cast</b>: Names of the cast for that movie or show
                <ul>
                    <li style="font-family:verdana;">Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane...</li>
                    <li style="font-family:verdana;">Antti Pääkkönen, Heljä Heikkinen, Lynne Guaglione, Pasi Ruohonen, Rauno Ahonen...</li>
                    <li style="font-family:verdana;">Luna Wedler, Jannis Niewöhner, Milan Peschel, Edin Hasanović...</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>country</b>: List of countries in which the content is available
                <ul>
                    <li style="font-family:verdana;">United States, South Africa, India...</li>
                    <li style="font-family:verdana;">United States, United Kingdom, Canada, Germany...</li>
                    <li style="font-family:verdana;">Canada, France, Japan, Russia...</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>date_added</b>: The date the content was added
                <ul>
                    <li style="font-family:verdana;">September 25, 2021</li>
                    <li style="font-family:verdana;">September 22, 2021</li>
                    <li style="font-family:verdana;">August 23, 2021</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>release_year</b>: Release year of the content
                <ul>
                    <li style="font-family:verdana;">2020</li>
                    <li style="font-family:verdana;">2014</li>
                    <li style="font-family:verdana;">2006</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>rating</b>: Motion picture content rating
                <ul>
                    <li style="font-family:verdana;">PG-13</li>
                    <li style="font-family:verdana;">TV-14</li>
                    <li style="font-family:verdana;">R</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>duration</b>: Total duration of the content
                <ul>
                    <li style="font-family:verdana;">90 min</li>
                    <li style="font-family:verdana;">2 seasons</li>
                    <li style="font-family:verdana;">48 min</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>listed_in</b>: List of categories where content is listed
                <ul>
                    <li style="font-family:verdana;">Documentaries, International TV Shows, TV Dramas, TV Mysteries...</li>
                    <li style="font-family:verdana;">Comedies, Dramas, International Movies...</li>
                    <li style="font-family:verdana;">Documentaries, TV Dramas, Comedies...</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
            <li style="font-family:verdana;"><b>description</b>: Short description about the content
                <ul>
                    <li style="font-family:verdana;">As her father nears the end of his life, filmmaker Kirsten Johnson stages his death in inventive and comical ways to help them both face the inevitable</li>
                    <li style="font-family:verdana;">After crossing paths at a party, a Cape Town teen sets out to prove whether a private-school swimming star is her sister who was abducted at birth</li>
                    <li style="font-family:verdana;">Tayo speeds into an adventure when his friends get kidnapped by evil magicians invading their city in search of a magical gemstone</li>
                    <li style="font-family:verdana;">...</li>
                </ul>
            </li>
        </ol>
    </div>
</body>

<div id="checking-null-values"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h3 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">3.1 Checking Null Values
    </h3>
</div>

<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Let's first check if there are any null values in the dataset.</p>
    </div>
</body>

In [3]:
df.isna().sum().sort_values(ascending=False)

director        2634
country          831
cast             825
date_added        10
rating             4
duration           3
show_id            0
type               0
title              0
release_year       0
listed_in          0
description        0
dtype: int64

<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Let's first check if there are any null values in the dataset.</p>
        <ul>
            <li style="font-family:verdana;">There are small amount of null values in the <b>date_added</b>, <b>rating</b>, <b>duration</b> columns.</li>
            <li style="font-family:verdana;">Meanwhile, there are large amount of null values in the <b>country</b>, <b>cast</b>, <b>director</b> columns.</li>
        </ul>
    </div>
</body>

<div id="optimizing-the-dataset"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h3 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">3.2 Optimizing the Dataset
    </h3>
</div>

<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-radius: 10px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Many of the features are <b>object</b> type. We can convert those features into more specific and compact types to save memory and increase efficiency.</p>
    </div>
</body>


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Let's first check the number of unique values for each feature to make the conversion process easier..</p>
    </div>
</body>

In [5]:
for i in df.columns:
    print(f"Number of unique values in the {i} column: {df[i].nunique()}")

Number of unique values in the show_id column: 8807
Number of unique values in the type column: 2
Number of unique values in the title column: 8807
Number of unique values in the director column: 4528
Number of unique values in the cast column: 7692
Number of unique values in the country column: 748
Number of unique values in the date_added column: 1767
Number of unique values in the release_year column: 74
Number of unique values in the rating column: 17
Number of unique values in the duration column: 220
Number of unique values in the listed_in column: 514
Number of unique values in the description column: 8775


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Some of the columns contain less unique values. These will be the ones that we do conversion on them. Columns that contain less unique values:</p>
        <ul>
            <li><b>type</b></li>
            <li><b>release_year</b></li>
            <li><b>rating</b></li>
        </ul>         
        <p style="font-family:verdana;">Other interesting features are <b>show_id</b>, <b>title</b>, and <b>description</b>. They contain unique values for each row. I will check them first to see what is going.</p>
    </div>
</body>



In [6]:
print("Number of unique values:", df["show_id"].nunique())
print(df["show_id"].unique())

Number of unique values: 8807
['s1' 's2' 's3' ... 's8805' 's8806' 's8807']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There is a unique value for each show. I don't understand the point of this feature. It only contains the values starting from 1 to 8807 which is basically an index for every row. The problem is that we already have that information stored in our dataframe so, I will remove this feature because it doesn't contain any valuable information for the analysis.</p>
    </div>
</body>

In [7]:
df.drop(labels="show_id", axis=1, inplace=True)

In [8]:
print("Number of unique values:", df["title"].nunique())
print(df["title"].unique())

Number of unique values: 8807
['Dick Johnson Is Dead' 'Blood & Water' 'Ganglands' ... 'Zombieland'
 'Zoom' 'Zubaan']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Similar to <b>show_id</b>, this column also contains unique values for each show but the difference is that these values are not just integers like an index value. These values contain the name of the each show which is probably will be used in the analysis. Therefore, I will not remove this feature.</p>
    </div>
</body>

In [9]:
print("Number of unique values:", df["description"].nunique())
print(df["description"].unique())

Number of unique values: 8775
['As her father nears the end of his life, filmmaker Kirsten Johnson stages his death in inventive and comical ways to help them both face the inevitable.'
 'After crossing paths at a party, a Cape Town teen sets out to prove whether a private-school swimming star is her sister who was abducted at birth.'
 'To protect his family from a powerful drug lord, skilled thief Mehdi and his expert team of robbers are pulled into a violent and deadly turf war.'
 ...
 'Looking to survive in a world taken over by zombies, a dorky college student teams with an urban roughneck and a pair of grifter sisters.'
 'Dragged from civilian life, a former superhero must train a new crop of youthful saviors when the military preps for an attack by a familiar villain.'
 "A scrappy but poor boy worms his way into a tycoon's dysfunctional family, while facing his fear of music and the truth about his past."]


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">Similar to <b>title</b>, this column also contains unique values for each show. These values contain the short description of the each show which we may use it in the analysis. Therefore, I will not remove this feature.</p>
    </div>
</body>

In [10]:
print("Number of unique values:", df["type"].nunique())
print(df["type"].unique())

Number of unique values: 2
['Movie' 'TV Show']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There are only two unique values in the <b>type</b> column. These are essentially used to categorize the content into <b>Movie</b> or <b>TV-Show</b>. I will convert this feature into <b>category</b> type.</p>
    </div>
</body>

In [11]:
df["type"] = df["type"].astype("category")

In [12]:
print("Number of unique values:", df["release_year"].nunique())
print(df["release_year"].unique())

Number of unique values: 74
[2020 2021 1993 2018 1996 1998 1997 2010 2013 2017 1975 1978 1983 1987
 2012 2001 2014 2002 2003 2004 2011 2008 2009 2007 2005 2006 1994 2015
 2019 2016 1982 1989 1990 1991 1999 1986 1992 1984 1980 1961 2000 1995
 1985 1976 1959 1988 1981 1972 1964 1945 1954 1979 1958 1956 1963 1970
 1973 1925 1974 1960 1966 1971 1962 1969 1977 1967 1968 1965 1946 1942
 1955 1944 1947 1943]


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There are 74 unique values in the <b>release_year</b> column. Also, these are integer values. Smallest integer datatype that can hold these values is <b>int16</b>.</p>
    </div>
</body>

In [13]:
df["release_year"] = df["release_year"].astype("int16")

In [14]:
print("Number of unique values:", df["rating"].nunique())
print(df["rating"].unique())

Number of unique values: 17
['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' nan 'TV-Y7-FV' 'UR']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There are 17 unique values in the <b>rating</b> column. These values contain both integers and strings. Therefore, we will convert this column into <b>category</b> type.</p>
    </div>
</body>

In [15]:
df["rating"] = df["rating"].astype("category")

In [16]:
print("Number of unique values:", df["duration"].nunique())
print(df.loc[:30, "duration"].unique())

Number of unique values: 220
['90 min' '2 Seasons' '1 Season' '91 min' '125 min' '9 Seasons' '104 min'
 '127 min' '4 Seasons' '67 min' '94 min' '5 Seasons' '161 min' '61 min'
 '166 min' '147 min' '103 min' '97 min' '106 min' '111 min']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There are 220 unique values in the <b>duration</b> column. Considering there are 8807 rows, there will be a lot of repetition. That's why I will convert it into <b>category</b> type.</p>
    </div>
</body>

In [17]:
df["duration"] = df["duration"].astype("category")

In [18]:
print("Number of unique values:", df["director"].nunique())
print(df.loc[:20, "director"].unique())

Number of unique values: 4528
['Kirsten Johnson' nan 'Julien Leclercq' 'Mike Flanagan'
 'Robert Cullen, José Luis Ucha' 'Haile Gerima' 'Andy Devonshire'
 'Theodore Melfi' 'Kongkiat Komesiri' 'Christian Schwochow'
 'Bruno Garotti' 'Pedro de Echave García, Pablo Azorín Williams'
 'Adam Salky' 'Olivier Megaton']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There are 4528 unique values in the <b>director</b> column. Similar to <b>duration</b> column, there are way too many unique values for the <b>category</b> type. I will not convert it into <b>category</b> type.</p>
    </div>
</body>

In [19]:
print("Number of unique values:", df["date_added"].nunique())
print(df.loc[:20, "date_added"].unique())

Number of unique values: 1767
['September 25, 2021' 'September 24, 2021' 'September 23, 2021'
 'September 22, 2021']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There are 1767 unique values in the <b>date_added</b> columns. We can use <b>datetime64</b> type for this column. This will helps us to sort dates easier.</p>
    </div>
</body>

In [20]:
df["date_added"] = df["date_added"].astype("datetime64")

In [21]:
print("Number of unique values:", df["country"].nunique())
print(df.loc[:20, "country"].unique())

Number of unique values: 748
['United States' 'South Africa' nan 'India'
 'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia'
 'United Kingdom' 'Germany, Czech Republic' 'Mexico']


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">There are 748 unique values in the <b>country</b> column. Because of the same reason with <b>duration</b> column, I will convert this into <b>category</b> type.</p>
    </div>
</body>

In [22]:
df["country"] = df["country"].astype("category")

In [23]:
print(df.loc[:5, "cast"].unique(), "\n")
print(df.loc[:5, "listed_in"].unique())

[nan
 'Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy Mahlangu, Ryle De Morny, Greteli Fincham, Sello Maake Ka-Ncube, Odwa Gwanya, Mekaila Mathys, Sandi Schultz, Duane Williams, Shamilla Miller, Patrick Mofokeng'
 'Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera'
 'Mayur More, Jitendra Kumar, Ranjan Raj, Alam Khan, Ahsaas Channa, Revathi Pillai, Urvi Singh, Arun Kumar'
 'Kate Siegel, Zach Gilford, Hamish Linklater, Henry Thomas, Kristin Lehman, Samantha Sloyan, Igby Rigney, Rahul Kohli, Annarah Cymone, Annabeth Gish, Alex Essoe, Rahul Abburi, Matt Biedel, Michael Trucco, Crystal Balint, Louis Oliver'] 

['Documentaries' 'International TV Shows, TV Dramas, TV Mysteries'
 'Crime TV Shows, International TV Shows, TV Action & Adventure'
 'Docuseries, Reality TV'
 'International TV Shows, Romant

<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <p></p>
        <p style="font-family:verdana;">The <b>cast</b> and <b>listed_in</b> columns does contain list of values for each show. We can't do any conversion on this.</p>
        <p style="font-family:verdana;">We're done with the optimization. Let's check the dataframe after the conversion.</p>
    </div>
</body>

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   type          8807 non-null   category      
 1   title         8807 non-null   object        
 2   director      6173 non-null   object        
 3   cast          7982 non-null   object        
 4   country       7976 non-null   category      
 5   date_added    8797 non-null   datetime64[ns]
 6   release_year  8807 non-null   int16         
 7   rating        8803 non-null   category      
 8   duration      8804 non-null   category      
 9   listed_in     8807 non-null   object        
 10  description   8807 non-null   object        
dtypes: category(4), datetime64[ns](1), int16(1), object(5)
memory usage: 514.4+ KB


<body>
    <div style="background-color: #FF6A74;
                border: 2px solid;
                border-radius: 10px;
                padding-top: 10px;
                padding-bottom: 10px;
                padding-right: 20px;
                padding-left: 20px;
                border-color: #FF1100;"> 
        <ul>
            <li style="font-family:verdana;">Before the optimization, dataset had a memory usage of <b>825.8+ KB</b>.</li>
            <li style="font-family:verdana;">After the optimization, dataset have a memory usage of <b>514.4+ KB</b>.</li>
            <li style="font-family:verdana;">We lowered the memory usage by almost <b>38%</b>.</li>
        </ul>
        <p style="font-family:verdana;">The <b>cast</b> and <b>listed_in</b> columns does contain list of values for each show. We can't do any conversion on this.</p>
        <p style="font-family:verdana;">We're done with the optimization. Let's check the dataframe after the conversion.</p>
    </div>
</body>

<div id="exploratory-data-analysis"
     style = "display: fill;
              border-radius: 10px;
              background-color: #E50914;">
    <h2 style = "padding: 15px; 
                 color: White;
                 text-align: left;
                 font-family: Trebuchet MS;">4. Exploratory Data Analysis
    </h2>
</div>

In [25]:
fig = px.bar(data_frame=df, x="rating", template="plotly_dark")
fig.show()